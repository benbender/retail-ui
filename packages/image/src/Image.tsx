import { Theme, useThemeCtx } from '@retail-ui/theme'
import clsx from 'clsx'
import * as React from 'react'

export type ImageFit = keyof Theme['ImageStyles']['fit']
export type ImageFitPosition = keyof Theme['ImageStyles']['fitPosition']

type ReactDivProps = React.HTMLAttributes<HTMLImageElement>
type Ref = HTMLImageElement

export type ImageProps = {
  src: string
  alt: string
  className?: string
  /** How the image object fits its bounds */
  fit?: ImageFit
  /* Positioning of the fitted image. */
  fitPosition?: ImageFitPosition
  /* Fix the width of image. Not responsive of screen width is smaller than image */
  isFixed?: boolean
  width?: React.CSSProperties['width']
  height?: React.CSSProperties['height']
}

export const Image = React.forwardRef<Ref, ReactDivProps & ImageProps>(
  (props, ref) => {
    const {
      children,
      className,
      alt,
      src,
      fit,
      fitPosition,
      isFixed,
      width,
      height,
    } = props

    const {
      theme: { ImageStyles },
    } = useThemeCtx()

    const cls = clsx(
      className,
      ImageStyles.base,
      isFixed && ImageStyles.fixed,
      fit && ImageStyles.fit[fit],
      fitPosition && ImageStyles.fitPosition[fitPosition],
    )

    return (
      <img
        className={cls}
        ref={ref}
        alt={alt}
        src={src}
        style={{
          width,
          height,
        }}
      >
        {children}
      </img>
    )
  },
)

Image.displayName = 'Image'
