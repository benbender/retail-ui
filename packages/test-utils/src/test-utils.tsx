import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
