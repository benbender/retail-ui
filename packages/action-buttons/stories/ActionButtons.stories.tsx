import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
