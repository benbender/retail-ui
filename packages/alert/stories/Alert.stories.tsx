import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
