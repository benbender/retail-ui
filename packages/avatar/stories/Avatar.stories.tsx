import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
