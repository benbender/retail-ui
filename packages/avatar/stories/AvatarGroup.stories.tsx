import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
