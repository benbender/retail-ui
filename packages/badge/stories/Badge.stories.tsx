import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
