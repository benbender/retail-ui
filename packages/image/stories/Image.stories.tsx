import * as React from 'react'

import { Image, ImageProps } from '../src'

const meta = {
  title: 'Retail-UI/Image',
  component: Image,
}

export default meta
const src =
  'https://images.unsplash.com/photo-1556740738-b6a63e27c4df?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=448&q=80'

const Template = (args: ImageProps) => (
  <Image
    {...args}
    data-testid={meta.title}
    src={src}
    alt="Woman paying for a purchase"
    height="300px"
  />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
