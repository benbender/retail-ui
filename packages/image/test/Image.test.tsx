import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Image from '../stories/Image.stories'

describe('Image', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Image.Simple src={Image.src} alt="Woman paying for a purchase" />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
