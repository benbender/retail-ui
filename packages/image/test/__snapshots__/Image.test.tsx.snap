// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correctly 1`] = `
<DocumentFragment>
  <img
    alt="Woman paying for a purchase"
    class="bg-gray-400 dark:bg-gray-700 w-full mx-auto"
    src="https://images.unsplash.com/photo-1556740738-b6a63e27c4df?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=448&q=80"
    style="height: 300px;"
  />
</DocumentFragment>
`;
